presto-main/src/main/java/com/facebook/presto/operator/scalar/UnicodeUtil.java

```diff
@@ -0,0 +1,212 @@
+            //
+            // Count bytes which are NOT the start of a code point
+            i32 = ((i32 & TOP_MASK32) >>> 7) & (~i32 >>> 6);
+            count += (i32 * ONE_MULTIPLIER32) >>> 24;
+
+            i += 4;
+        }
+        //
+        // Do the rest one by one
+        for (; i < end; i++) {
+            int i8 = string.getByte(i) & 0xff;
+            //
+            // Count bytes which are NOT the start of a code point
+            count += (i8 >>> 7) & (~i8 >>> 6);
+        }
+
+        assert count <= end;
+        return end - count;
+    }
+
+    /**
+     * Finds the index of the first byte of the code point at a position.
+     */
+    static int findUtf8IndexOfCodePointPosition(final Slice string, final int codePointPosition)
+    {
+        //
+        // Quick exit if we are sure that the position is after the end
+        if (string.length() <= codePointPosition) {
+            return string.length();
+        }
+
+        int correctIndex = codePointPosition;
+        int i = 0;
+        //
+        // Length rounded to 8 bytes
+        final int length8 = string.length() & 0x7FFFFFF8;
+        //
+        // While we have enough bytes left and we need at least 8 characters process 8 bytes at once
+        while (i < length8 && correctIndex >= i + 8) {
+            //
+            // Fetch 8 bytes as long
+            long i64 = string.getLong(i);
+            //
+            // Count bytes which are NOT the start of a code point
+            i64 = ((i64 & TOP_MASK64) >>> 7) & (~i64 >>> 6);
+            correctIndex += ((i64 * ONE_MULTIPLIER64) >>> 56);
+
+            i += 8;
+        }
+        //
+        // Length rounded to 4 bytes
+        final int length4 = string.length() & 0x7FFFFFFC;
+        //
+        // While we have enough bytes left and we need at least 4 characters process 4 bytes at once
+        while (i < length4 && correctIndex >= i + 4) {
+            //
+            // Fetch 4 bytes as integer
+            int i32 = string.getInt(i);
+            //
+            // Count bytes which are NOT the start of a code point
+            i32 = ((i32 & TOP_MASK32) >>> 7) & (~i32 >>> 6);
+            correctIndex += ((i32 * ONE_MULTIPLIER32) >>> 24);
+
+            i += 4;
+        }
+        //
+        // Do the rest one by one, always check the last byte to find the end of the code point
+        while (i < string.length()) {
+            int i8 = string.getByte(i) & 0xff;
+            //
+            // Count bytes which are NOT the start of a code point
+            correctIndex += ((i8 >>> 7) & (~i8 >>> 6));
+            if (i == correctIndex) {
+                return i;
+            }
+
+            i++;
+        }
+
+        assert i == string.length();
+        return string.length();
+    }
+
+    /**
+     * Find substring within UTF-8 encoding string.
+     */
+    static int findUtf8IndexOfString(final Slice string, final int start, final int end, final Slice substring)
```

Comment:
I wanted to avoid the allocation of an additional buffer for each `indexOf`.
I looking forward to the results of the benchmarks with and without your `indexOf`.